POS Cart redesign

dennisreimann opened the pull request:
Closes #5109. Closes #5130.

Quite a few things happened here:

- Updated the POS cart UI, as this was the purpose of the issue
- Removed the custom amount input for the cart version (as discussed in the dev and design meetings)
- Removed the modal confirmation step from the cart, as the initial overview is sufficient
- Refactored the POS Cart from jQuery to Vue.js
- Extracted a Vue mixin with common functionality which is now used by the keypad and cart
- Unified all POS views as much as possible, especially the item lists

I can imagine that we have another round of more code finetuning for the POS views and CSS, but that would be covered in a separate PR then.

## Item List

![grafik](https://github.com/btcpayserver/btcpayserver/assets/886/90d94808-0208-437a-bef5-e31930538612)

## Cart

![grafik](https://github.com/btcpayserver/btcpayserver/assets/886/64604345-e002-4058-8ea6-40ac74df9b18)

dstrukt commented:
HELL YEAH. Can't wait to review!

Kevyke commented:
What is the site about

dennisreimann commented:
Adapted it so that the cart is displayed as an offcanvas for mobile (see screenshots above) and sidebar for desktop (screenshots below).

![dark](https://github.com/btcpayserver/btcpayserver/assets/886/1e3c926d-7f89-4c30-a80f-e6951b8dd254)

![light](https://github.com/btcpayserver/btcpayserver/assets/886/0e687154-2cac-4cec-a817-76c75876f596)

dstrukt reviewed (changes requested):
## Still Reviewing, will remove when done.

## Overall, this is fantastic, so much better!!

Some of the easy adjustments I'll do / can help with, but documenting for both of us:

## List View

- [x] The title of the list and cart view should be a consistent font-size, looked and it seems to use the `<h1 class="store-name">` instead..
- [ ] Should we add search/categories to the List view?

<img width="1340" alt="Screen Shot 2023-07-20 at 1 43 50 PM"  src="https://github.com/btcpayserver/btcpayserver/assets/6250771/1c37924b-49e2-46ca-8b8f-21d61b6aeb8b">

- [ ] Should we ellipses the content after X characters to prevent crazy differences in card height? It's fine here because the heights are pretty consistent, but could see it being an issue.
<img width="1350" alt="Screen Shot 2023-07-20 at 1 56 20 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/000e9a63-603d-41dd-a6dc-48f5e2b6861b">

## Cart View

### Desktop

- [x] We prevent the user from clicking past the inventory, but if you input greater than the inventory, shouldn't it just update back to the max? e.g. i'm able to input 8 with a 7 max, and i'm sure i'll hit an error, but it could be avoided with some JS, no?
<img width="418" alt="Screen Shot 2023-07-20 at 1 50 03 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/7edb21be-527e-4ed3-ac14-c42e3ea68942">

- [x] We might want to increase the width a little bit on the cart view, so we don't have wrapping issue? As much as I love 420 😎 maybe a bit more, 480? Open to other ideas here.
<img width="426" alt="Screen Shot 2023-07-20 at 1 48 03 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/3678755c-4e39-4ad2-8916-2c09daa90003">

- [x] Think we could just go with `All` vs. `All items` to save a little bit of horizontal space imo
<img width="1292" alt="Screen Shot 2023-07-20 at 1 47 43 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/67920256-9150-4651-a2af-299508efe7c6">

- [x] We should probably reduce the size of the cart icon from 40/40 to 32/32px, it's a bit large atm

- [x] We should also bump the badge font-size as well, it's currently too small, ideally 12pt or like `0.857...?em` or whatever the math is based on the default 14 pixel size

- [x] I'd love to be able to override the styling on these, the browser defaults are ugly
<img width="54" alt="Screen Shot 2023-07-20 at 2 04 42 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/e4693f17-bd38-44f5-b366-2f2940454c32">

- [x] Reduce the icon size to 2.25em from 2.5em
<img width="491" alt="Screen Shot 2023-07-20 at 2 30 01 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/a49693a7-ba21-4e00-8aae-be87c3bf2eb8">

- [ ] There is currently no way to close to the cart on desktop

- [x] I manage to break the number display when there is No quantity in the input, e.g: not sure if this requires action, but a note:
<img width="539" alt="Screen Shot 2023-07-20 at 2 11 05 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/a9790459-42ba-4ef5-b333-35c71358151f">

- [ ] I changed the title halfway through editing, but it didn't update in the cart view, minor edge case, but an interesting one:
<img width="1772" alt="Screen Shot 2023-07-20 at 2 17 22 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/2f0f5c93-e591-4f80-959a-0ffcc32bcc38">

- [x] Not 100% certain the table headers are necessary, so just thought I'd comment:
<img width="420" alt="Screen Shot 2023-07-20 at 2 09 02 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/2f1a9cbd-4730-4172-ab3c-c1492e863e54">

- [ ] Similarly, I've removed the magnifying glass from some of the search inputs, but debating if it makes sense to add them back...

- [x] Update placeholder text from `Find product` to `Search...` or `Search for product...`. Thoughts?

- [x] We should bump right and left the padding on the inputs:
<img width="1385" alt="Screen Shot 2023-07-20 at 3 10 44 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/5ac67a7d-cbdf-4c59-9683-b18624cc3374">

- [x] Debating if we want to round the border radius to full on the input for this view specifically. It's a bit inconsistent from all inputs, but think it looks nice and differentiates search on specific views. Thoughts?
<img width="1318" alt="Screen Shot 2023-07-20 at 3 10 59 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/5efc50c5-96d5-4c48-9fb4-2802dcc59387">

- [x] Also still trying to decide if I like the `Empty Cart` at the bottom of the view, logically it does make some sense, but it could accidentally be pressed and be quite annoying. I had originally placed it to the right of the logo / title, which would work on desktop in the current form, but might be a bit messy on mobile.. hmmm

- [ ] From a functional standpoint, clicking both the card and the button adds the product to the cart, and this is a minor functional nit .. but should it just be limited to the button to add things to the cart?

### Mobile

- [x] The `X` icon is black, not white
- [x] Should ideally be horizontally aligned with the title of the store as well
<img width="497" alt="Screen Shot 2023-07-20 at 2 06 23 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/d0af9621-e0b7-49b2-b02b-16384ba7cb0e">

- [ ] Should we make the pills scrollable like we do with the tabbed navigation? Thoughts?
<img width="493" alt="Screen Shot 2023-07-20 at 3 24 55 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/ce0ecab1-5904-4da8-8006-38b7d78c8005">

- [ ] Should there be an "empty cart" state? Currently you just can't click into the cart if it's empty. Thoughts? Am able to get this view when clearing the cart and degrading to mobile though - this works:
<img width="1095" alt="Screen Shot 2023-07-20 at 3 34 00 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/94089a13-7976-40ae-9f2a-24715299bcf1">

- [ ] Not 100% sure I like the slide-in animation for the cart on mobile, maybe just a dissolve in/out might be better. Thoughts? Side thought, but doing a 100% width view for the main navigation (as opposed to the the current 75-80%), and the animation would be treated the same, in either case.

## Additional Thoughts & Considerations:

- Do we want to include Store branding, logos, etc..? This might be a follow up PR, but it's something I could understand some users wanting.

dennisreimann commented:
Thanks @dstrukt, I addressed most of the feedback, see the recent commits.

> Should we add search/categories to the List view?

This one does not have any JavaScript yet and isn't as dynamic as the cart, so that would be a separate overhaul I'd say.

>  Should we ellipses the content after X characters to prevent crazy differences in card height? It's fine here because the heights are pretty consistent, but could see it being an issue.

This behaves as before and so far this hasn't been brought up, so I'd say we are good for this PR. However, I think this is a valid point to address in a future update.

>  I'd love to be able to override the styling on these, the browser defaults are ugly

Let's do this in an additional round of updates as well, as I'm time constrained for the rest of this release.

> From a functional standpoint, clicking both the card and the button adds the product to the cart, and this is a minor functional nit .. but should it just be limited to the button to add things to the cart?

@pavlenex and I spoke about this and decided to keep the current behaviour to not change the flow for merchants. It's easier to click the whole item than just the button, which makes it better to handle on mobile/tablets.

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
+            Assert.Contains("10% = 0,30 €", s.Driver.FindElement(By.Id("CartDiscount")).Text);
+            Assert.Equal("2,70 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
             
-            // Custom amount
-            s.Driver.FindElement(By.Id("CartCustomAmount")).SendKeys("1.5");
-            s.Driver.FindElement(By.Id("CartTotal")).Click();
-            Assert.Equal("3,50 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
-            s.Driver.FindElement(By.Id("js-cart-confirm")).Click();
```
Custom amount was emoved from the cart view as discussed.

dennisreimann commented on BTCPayServer/Views/Shared/PosData.cshtml:
```diff
@@ -23,8 +23,7 @@
                     }
-                    <td style="white-space:pre-wrap">
-                        @if (IsValidURL(str))
+                    <td style="white-space:pre-wrap">@* Explicitely remove whitespace at front here *@@if (IsValidURL(str))
```
Otherwise the value column starts with an empty line and the data is shown from line 2 on.

dennisreimann commented on BTCPayServer/Views/UIOnChainAutomatedPayoutProcessors/Configure.cshtml:
```diff
@@ -26,23 +26,23 @@
                 <label asp-for="IntervalMinutes" class="form-label" data-required>Interval</label>
                 <div class="input-group">
-                    <input asp-for="IntervalMinutes" class="form-control" inputmode="numeric" style="max-width:10ch;">
+                    <input asp-for="IntervalMinutes" class="form-control" inputmode="numeric" style="max-width:12ch;">
```
These changes needed to be done as we bumped the horizontal padding of the input fields in the design system.

dennisreimann commented:
Updated the cart display to incorporate @dstrukt's feedback/

![dark](https://github.com/btcpayserver/btcpayserver/assets/886/522019a8-5f10-44cc-b607-460fc087d828)

![light](https://github.com/btcpayserver/btcpayserver/assets/886/31c9f67d-0cc5-4d28-83e1-906b181871b4)

pavlenex reviewed (approved):
tACK, great work @dennisreimann and thanks @dstrukt for in-depth testing and feedback.

It looks great. My only comment is that on destkop view the icon cart still remains though it doesn't really do much. I like how this was handled for mobile and in generally very aesthetically pleasing. Good work team!

dennisreimann commented:
I thought it's nice to have it with the badge, but we can obviously also remove it on desktop. Let us know what you think, @dstrukt

dstrukt commented:
I see both sides, and agree with @dennisreimann that it looks nice with the amount badge, but think I expect it to open/close the cart as well, because that's how it functions on mobile.

Hmm, the lack of functionality on desktop doesn't bother me, and think we should just keep it for the time being / see if get any additional feedback that it's annoying or confusing.

Will do a proper round of review on the last set of changes, thanks for knocking these out so quickly @dennisreimann !!

dstrukt commented:
> This one does not have any JavaScript yet and isn't as dynamic as the cart, so that would be a separate overhaul I'd say.

Sounds good, and agree, totally finally to leave out on this, wasn't sure how far the scope went. 

> This behaves as before and so far this hasn't been brought up, so I'd say we are good for this PR. However, I think this is a valid point to address in a future update.

Sounds good as well, an easy enough patch if the issue arises with more usage.
 
> Let's do this in an additional round of updates as well, as I'm time constrained for the rest of this release.

Sounds good, there's definitely some very minor design system stuff that I need to make a few issues for.
 
> @pavlenex and I spoke about this and decided to keep the current behaviour to not change the flow for merchants. It's easier to click the whole item than just the button, which makes it better to handle on mobile/tablets.

> We also talked about this and decided to leave out this functionality.

Sounds good wrt the former, will see if the lack of functionality on the icon, as discussed above, will be an issue.

dstrukt commented:
## Final thoughts:

Overall, love the update, looks fantastic - there are a few design system nits that will get addressed in after this is merged, as well as improving the singular list view, so that will all be wrapped up nicely.

Last few nits, don't need them to be done, but if they are, great, and won't require a follow up review.

- [ ] Empty button should probably be green
<img width="194" alt="Screen Shot 2023-07-21 at 12 02 34 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/ced31ac8-ed1f-40f2-9570-1f2e22da76d7">

- [ ] Thought I made this change, but should be `All`, unless I misunderstood something?
<img width="239" alt="Screen Shot 2023-07-21 at 12 02 29 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/0c4e3fde-dc74-4111-9691-dd41c3c4ba38">

- [ ] Might be a separate follow up, but this button hover state should be a shade or two lighter than the default
<img width="218" alt="Screen Shot 2023-07-21 at 12 02 22 PM" src="https://github.com/btcpayserver/btcpayserver/assets/6250771/560c3d10-8fbd-49fc-8d7c-4a17a95147be">

dstrukt reviewed (approved):
LEGGO!

dstrukt commented:
If we remove the cart icon, don't we want to scale up the badge to 32px, and have a "0" state? Just saw the latest commits.

pavlenex commented:
Alright, seems we can merge this one then @NicolasDorier 🚀

dstrukt commented:
Ahh I see, it's just removed. LGTM!

dennisreimann commented on BTCPayServer/Views/Shared/PointOfSale/Public/Cart.cshtml:
```diff
@@ -1,349 +1,244 @@
+            <div class="container-xl">
+                <header class="sticky-top bg-tile offcanvas-header py-3 py-lg-4 d-flex align-items-baseline justify-content-center gap-3 px-5 pe-lg-0">
+                    <h1 class="mb-0" id="cartLabel">Cart</h1>
+                    <button id="CartClear" type="reset" v-on:click="clearCart" class="btn btn-text text-primary p-1">
```
```suggestion
                    <button id="CartClear" type="reset" v-on:click="clearCart" class="btn btn-text text-primary p-1" v-if="cartCount > 0">
```

dennisreimann commented:
> Visual edge case to have the Empty button on ana empty cart though. Minor.

Fixed.

dstrukt commented:
LGTM! Ready for merge.

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
-            Assert.Single(s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")));
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(1) .btn-primary")).Click();
+            Assert.Single(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
+            Assert.Equal("1", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
```suggestion
```

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
-            Assert.Empty(s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")));
             Thread.Sleep(250);
+            Assert.Empty(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
```
```suggestion
```

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
-            Assert.Equal("0,00 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
-            Assert.False(s.Driver.FindElement(By.Id("CartClear")).Displayed);
+            s.Driver.WaitForElement(By.Id("PosItems"));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
```
```suggestion
```

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(2) .btn-primary")).Click();
             Thread.Sleep(250);
-            s.Driver.FindElement(By.CssSelector(".card.js-add-cart:nth-child(1)")).Click();
+            Assert.Equal("1", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
```suggestion
```

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
             Thread.Sleep(250);
-            Assert.Equal(2, s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")).Count);
-            Assert.Equal("2,00 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
+            Assert.Equal("2", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
```suggestion
```

dennisreimann commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
+            Assert.Equal("2", s.Driver.FindElement(By.Id("CartBadge")).Text);
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(2) .btn-primary")).Click();
+            Thread.Sleep(250);
+            Assert.Equal("3", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
```suggestion
```
